lnst/RecipeCommon/Perf/Measurements/LinuxPerfMeasurement.py

```diff
@@ -0,0 +1,241 @@
+            logging.error(f"timeout while waiting for perf job to finish")
+
+
+class LinuxPerfMeasurement(BaseMeasurement):
+    _MEASUREMENT_VERSION: int = 1
+
+    hosts: List[Host]
+    intr_filename: str
+    intr_cpus: List[int]
+    iperf_filename: str
+    iperf_cpus: List[int]
+    _start_timestamp: float
+    _end_timestamp: float
+
+    _version: Optional[Dict[str, Any]] = None
+    _collection_index: int = 0
+
+    def __init__(
+        self,
+        hosts: List[Host],
+        intr_filename: str,
+        intr_cpus: List[int],
+        iperf_filename: str,
+        iperf_cpus: List[int],
+        recipe_conf: EnrtConfiguration = None,
+    ):
+        super().__init__(recipe_conf)
+
+        self.hosts = hosts
+        self.intr_filename = intr_filename
+        self.intr_cpus = intr_cpus
+        self.iperf_filename = iperf_filename
+        self.iperf_cpus = iperf_cpus
+
+        self.configurations = [
+            LinuxPerfMeasurementConfiguration(host, filename, cpu)
+            for host, filename, cpu in self._generate_configurations()
+        ]
+
+    @property
+    def version(self) -> Optional[Dict[str, Any]]:
+        if self._version:
+            return self._version
+
+        perf_version: Optional[str] = self._get_host_perf_version()
+        if perf_version is None:
+            return None
+
+        self._version = {
+            "measurement_version": self._MEASUREMENT_VERSION,
+            "host_perf_version": perf_version,
+        }
+        return self._version
+
+    def _get_host_perf_version(self) -> Optional[str]:
+        if not len(self.hosts):
+            raise Exception(
+                "No hosts in LinuxPerfMeasurement while getting perf version"
+            )
+
+        version_job = self.hosts[0].run("perf --version", job_level=ResultLevel.DEBUG)
+        if version_job.passed:
+            match = re.match(r"perf version (.+?)", version_job.stdout)
+            if match:
+                return match.group(1)
+        return None
+
+    def _generate_configurations(self):
+        for host in self.hosts:
+            yield (host, self.intr_filename, self.intr_cpus)
+            yield (host, self.iperf_filename, self.iperf_cpus)
+
+    def start(self) -> None:
+        self._start_timestamp = time.time()
+        for cfg in self.configurations:
+            cfg.run()
+
+    def finish(self) -> None:
+        for cfg in self.configurations:
+            cfg.kill()
+        self._end_timestamp = time.time()
+
+    def collect_results(self) -> List[BaseMeasurementResults]:
+        self._collection_index += 1
+        results: List[BaseMeasurementResults] = []
+        for cfg in self.configurations:
+            if cfg.job is None or cfg.job.result is None:
+                continue
+
+            # copy agent's perf.data file to a controller
+            perf_folder_path: str = os.path.abspath("linuxperf-data")
+            src_filepath: str = os.path.abspath(cfg.job.result["filename"])
```

Comment:
careful about the use of `os.path.abspath` here as this is running in the **Controller** process, while the `cfg.job.result["filename"]` was generated in the **Agent** process context which may be running in a different working directory, imo this call should be moved to the test module itself.

Reply:
yes, this was a mistake. I already called `os.path.abspath` in the Agent, but threw it in once more here for some reason. It's obviously not needed nor wanted here